### ISSUE TITLE
Offloading recordings to an external server

### ISSUE THREAD

**Issue by matiasilva** [bigbluebutton/bigbluebutton#9708]
**Is your feature request related to a problem? Please describe.**
At present, recording processing takes places on the same machine as the BBB components that manage video conferencing. This means that additional strain is placed on the CPU.

**Describe the solution you'd like**
The ability to offload recordings to a special server that is dedicated for recording processing. I managed to do this, but in a very hacky way that disables certain bbb-rap-X workers on the BBB hosts and activates others on the recording processing machine. Though it works, this breaks the bbb-record command.

**Describe alternatives you've considered**
Another idea would be to schedule recording processing to happen at night, when nobody is using the platform.

**Comment by iSamof** [comment]
I have seen other discussions about optimising the processing to leave more resources for live meetings. 

But an external processing server is a great idea. It is something we have considered but so far not attempted due the complications you have stated. For now, we have taken the safer route of increasing servers as demand rises. That is not sustainable and a waste of resources. 

So scheme to use a combination of an external processing server(s) plus load balancing to conferencing servers when not in use (or in low usage) would be excellent use of resources.

**Comment by matiasilva** [comment]
On our deployment of BBB, we have the recording pipeline fully working now. The files are dumped to a shared NFS mount, so the BBB hosts themselves have only 20GB of disk. They are then archived and run the sanity checks, at which point they are picked up by the external recording processing server. They then undergo processing and are transferred to the Scalelite directory to be served on Greenlight. Works perfectly without much overhead and without any drastic changes to BBB.

**Comment by stefantruehl** [comment]
Recently we released an alternative BigBlueButton frontend/scaling solution called BBB@Scale. Basically it is combining the features of Greenlite and Scalelite (not using these but reimplementing the features) + some additional stuff. Finde infos here: https://gitlab.com/bbbatscale/bbbatscale 

In a future release we will also release our recording management functionality. We have finished it and it is running in our installation, however, we were not able to document and clean it up for our 1.0 release. 

The feature allows you to define dedicated postprocessing and/or viewing BBB servers that are not used for regular meetings. So you can do special vHW configurations for the different purposes (postprocessing; viewing; hosting meetings). We found that handy for our installation at our university. 

So if this is what you are looking for, feel free to follow the project

**Comment by matiasilva** [comment]
@stefantruehl yes, I've been following the dev mailing list :) Thanks again for sharing, would love to find out more about how it works too!

**Comment by ryprfpryr** [comment]
Hey @matiasilva any update on this? I think this enhancement would set the next level of BBB if included out-of-the box.

For BBB servers running behind a Scalelite setup this is totally necessary for efficiency matters, and as far I can understand it should would work similarly like now but it should transfer the RAW recordings files in order to process them on another BBB server (set up just for recordings processing). When the recording is done, the post_process script should transfer them to Scalelite. Am I right?

Did you make it work?

**Comment by matiasilva** [comment]
I haven't been actively working on improving the way this is done in BBB. I have my own personal setup that Just Works. I expect that as 2.3 comes out I'll rework things then.

**Comment by GhaziTriki** [comment]
Dear @matiasilva, there is multiple ways to off-laod the recordings from a BigBlueutton server:
1 - Use a post-publish script.
2 - Mount a drive or block storage on /var/bigbluebutton

Both solutions are working on production. I think we can close this issue. I will let you decide @matiasilva

**Comment by pbdco** [comment]
> Dear @matiasilva, there is multiple ways to off-laod the recordings from a BigBlueutton server:
> 1 - Use a post-publish script.
> 2 - Mount a drive or block storage on /var/bigbluebutton
> 
> Both solutions are working on production. I think we can close this issue. I will let you decide @matiasilva

I think the purpose of this issue is to have the ability to process the recordings on a different server than the BBB server on which the meeting has been processed, BEFORE it is processed, and NOT after.

**Comment by GhaziTriki** [comment]
Then just use `bbb-record --disable` and use a `post_archive` script. Then you will be able to process them in another server.

**Comment by pbdco** [comment]
> Then just use `bbb-record --disable` and use a `post_archive` script. Then you will be able to process them in another server.

Thanks for the clue @GhaziTriki 
@matiasilva any suggestions on the post_archive script?

**Comment by GhaziTriki** [comment]
@pbdco Here is an example https://github.com/bigbluebutton/bigbluebutton/blob/develop/record-and-playback/core/scripts/post_archive/post_archive.rb.example

You can use RClone tool to offload your recording. It would be great to move this discussion to the forums later.
